### PR TITLE
enable tokio_unstable broadly

### PR DESCRIPTION
**Pull request by guswynn**
This pr is a followup to incident 70 that broadly enables `--cfg=tokio_console` and the `tokio` `tracing` feature across the codebase. This enables two main things:

- Allows us to use the `tokio-console` in production without special builds
- Allows us to expose runtime-level metrics from the runtime, which we can eventually expose to prometheus

in addition to various other nice lil apis, like `JoinSet` and `Handle::id`

I did an audit of the tokio codebase to see what this `cfg` and the `tracing` feature actually changed about the crate, to ensure that nothing was something unsafe. I found:

- Various additional apis that could break in the future.
  - We can deal with these easily when we upgrade tokio versions
- A new alternative runtime that we won't use.
- Some additional runtime options that we MAY want to try in the future, like `disable_lifo` and deterministic rng seeding
- **Runtime-level metrics**
  - As far as I can tell, at least some of these are already collected but not exposed, which means the cpu/memory cost should not be a problem. Some more expensive ones are off by default.
- The ability to dump stack traces of tasks. This is behind an additional `--cfg` flag as it may be expensive and very unstable. Might be neat in the future.
- `tracing` `Span`s attached to tasks and MOST `sync` resources

The main concerns here are the cpu/memory cost of metric-collection and all the new tracing. The metrics don't appear expensive, and with a brief look, it seemed like most of it was already collected. The cpu/memory cost of the tracing spans is hard to quantify, but a feature-benchmark run on an earlier version passed.

The other concern is related to the second commit. The commit has more information, but basically, tokio adding additional hidden spans to effectively all sync resources (and tasks) can cause bugs when multiple `Dispatch`s exist in the same process. This only happens in two locations, which now have scary comments and are audited. 

These new tokio span's also may increase the chance that we hit the bug as described in the original version of https://github.com/MaterializeInc/materialize/pull/20706, but the tracing fast-path test does pass on this pr, so I am inclined to move forward with it. I don't have proof or a complete understanding of that bug, so this is just speculation (cc @pH14 @danhhz)

### Motivation

  * This PR adds a known-desirable feature.

### Tips for reviewer
see above

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by guswynn** [comment]
Nightlies seem to pass on this, except `Product Limits` failed at least once. Look transient, re-trying. I also started a Release Qualification build: https://buildkite.com/materialize/release-qualification/builds/338

File: src/adapter/src/coord/sequencer/inner.rs [hunk 1]
```diff
@@ -3181,6 +3191,30 @@ impl Coordinator {
         Ok(Self::send_immediate_rows(rows))
     }
 
+    /// WARNING, ENTERING SPOOKY ZONE 3.0
```
Comment (guswynn) [inline src/adapter/src/coord/sequencer/inner.rs:3194]: @aalexandrov note the new subtlety here; If we switch to the fixed "broken" impl, I believe this goes away

do you have a draft branch for fixing that? I might be able to take a look

**Comment by guswynn** [comment]
Oh, the product limits error may be a stack overflow, ill have to look into that

**Comment by guswynn** [comment]
Increased the coordinator stack size again; I didn't dig into exactly what put us over the edge, but additional tokio span creation could cause it

**Comment by aalexandrov** [comment]
> Increased the coordinator stack size again.

I had to do a similar thing yesterday in #21973, although I thought the reason is because I moved around some optimization stages from the `plan_~` to the `sequence_~` functions.

File: src/adapter/src/coord/sequencer/inner.rs (hunk 1 above)
Comment (aalexandrov) [inline src/adapter/src/coord/sequencer/inner.rs:3194]: BTW, this is not the only such method. Look for `~_optimizer_pipeline` methods in the same file.

File: src/adapter/src/coord/sequencer/inner.rs [hunk 2]
```diff
@@ -3181,6 +3191,30 @@ impl Coordinator {
...
+    /// the thread-local default `Dispatch` using `.with_subscriber` at call-sites.
+    ///
+    /// In general, this should be considered safe, but `tracing` has limitations that mean that
+    /// any `Span` created under the specialized `Subscriber` that _leaves_ this function will
+    /// almost assuredly cause a panic inside `tracing`. This is because `Span`s track the
+    /// `Subscriber` they were created under, but certain actions (like an ordinary `Span` exit)
+    /// will call a method on the _thread-local_ `Subscriber`, which may be backed with a different
+    /// `Registry`.
+    ///
+    /// At first glance, there is no obvious way this method leaks `Span`s, but ANY tokio
```
Comment (aalexandrov) [inline src/adapter/src/coord/sequencer/inner.rs:3206]: OK, so IIUC the invariant is that any `.async` methods nested inside `~_optimizer_pipeline` methods need to be called with a `.with_subscriber(root_dispatch)` using a specifically passed `rood_dispatch`. BTW none of the other `~_optimizer_pipeline` has nested `.await` calls.

**Comment by aalexandrov** [comment]
> Increased the coordinator stack size again; I didn't dig into exactly what put us over the edge, but additional tokio span creation could cause it

@guswynn take a look [at this diff](https://github.com/MaterializeInc/materialize/commit/85e4464ac606f6dd5c728f795db04605ba80b599#diff-a3ef7f189b5ce8bb9265d229a8978309ecf36b29c9c6f7f3071950103d545289). I reduced the value from the default (1000) to (950) to make the nightly build happy yesterday. 

After the upgrade to the new tokio console it could be that we need to restrict this a bit more.

File: src/adapter/src/coord/sequencer/inner.rs (hunk 1 above)
Comment (aalexandrov) [reply src/adapter/src/coord/sequencer/inner.rs:3194]: I don't have a draft branch. I went as far as trying to save the `STACK: OnceLock<Arct<dyn Subscriber + Send + Sync>>` before calling `.init()`. See my comment in #21809.

Comment (guswynn) [reply src/adapter/src/coord/sequencer/inner.rs:3194]: sounds good, will look into this!

**Comment by guswynn** [comment]
Product Limits failed again, but not from a stack overflow, but retrying fixed it

File: src/adapter/src/coord/sequencer/inner.rs (hunk 2 above)
Comment (guswynn) [reply src/adapter/src/coord/sequencer/inner.rs:3206]: yes!

File: misc/python/materialize/cli/run.py [hunk 3]
```diff
@@ -305,9 +305,6 @@ def _build(args: argparse.Namespace, extra_programs: list[str] = []) -> int:
     env = dict(os.environ)
     command = _cargo_command(args, "build")
     features = []
-    if args.tokio_console:
-        features += ["tokio-console"]
-        env["RUSTFLAGS"] = env.get("RUSTFLAGS", "") + " --cfg=tokio_unstable"
```
Comment (benesch) [inline misc/python/materialize/cli/run.py:310]: Excited to see this go. You missed a spot though!

https://github.com/MaterializeInc/materialize/blob/eea35ba4ae1712159d7d20f36c5b6f81965b01f2/misc/python/materialize/cli/run.py#L103-L107

**Review by benesch — APPROVED** [review]

**Review by ParkMyCar — APPROVED** [review]

**Comment by guswynn** [comment]
This need to be rebased, as the `cargo-test` failure should be fixed on main, but the failing feature-benchmark (`InsertMultiRow`) is a known regression

File: misc/python/materialize/cli/run.py (hunk 3 above)
Comment (guswynn) [reply misc/python/materialize/cli/run.py:310]: I am leaving this so people don't need to know to pass `--tokio-console-listen-addr=127.0.0.1:6669`, but can remove in followup if need be

Comment (benesch) [reply misc/python/materialize/cli/run.py:310]: Oh, I think we should enable that listen address by default. No reason not to, right?

Comment (guswynn) [reply misc/python/materialize/cli/run.py:310]: It turns off the tracing fast-path optimization as tokio-console traces are `trace!` level, and may use an arbitrary amount of memory, is that considered reasonable as the default?

**Comment by danhhz** [comment]
Might be causing this on main?

```
error[E0432]: unresolved import `crate::netio`
  --> src/ore/src/tracing.rs:56:12
   |
56 | use crate::netio::SocketAddr;
   |            ^^^^^ could not find `netio` in the crate root
```

Makes me wonder, now that we have workspace-hack are the cfg features in mz-ore getting us much?

**Comment by guswynn** [comment]
> Makes me wonder, now that we have workspace-hack are the cfg features in mz-ore getting us much?


yeah probably

File: misc/python/materialize/cli/run.py (hunk 3 above)
Comment (benesch) [reply misc/python/materialize/cli/run.py:310]: Ah, got it, probably good to leave it disabled by default then.

**Pull request merged**